### PR TITLE
Kmer Overhaul

### PR DESCRIPTION
**Pull request by bicycle1885**
DONE:

* Moved k-mer related types and functions to kmer.jl.
* Nucleotides in a k-mer are filled from the MSBs to the LSBs and right-aligned.
* Fixed related methods to follow this layout change (except `EachKmerIterator`).
    * Given A < C < G < T, k-mer is ordered by the lexicographical order (ref #120).

TODO:

* [x] Fix `EachKmerIterator`.
* [x] Check performance regression (but in my limited micro benchmarks, I couldn't detect that.)

### CODE REVIEW

**Comment by bicycle1885** [comment]
Now ready to review.

Summary:

* Move k-mer related types and methods to two files: `kmer.jl` and `eachkmer.jl`.
* Nucleotides in a k-mer are filled from the MSBs to the LSBs and right-aligned.
    * This will lexicographically order k-mers (assume A < C < G < T),
    * and the next k-mer of a k-mer can be generated by `DNAKmer{4}(UInt64(kmer) + 1)`.
* Rewrite the algorithms of `EachKmerIteartor`.
    * Now (I hope) these are easier to read (see `eachkmer.jl`).
* Generate a `GeneticCode` object by parsing a code table.
    * This code table is available at: http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=tgencodes#SG1.

**Comment by bicycle1885** [comment]
I've tried a simple benchmark to check the performance of the k-mer sliding window:

```julia
using Bio.Seq

function bench_kmercounts(seq, K, step)
    println("DNAKmerCounts{$K}(seq)")
    println("step = $step")
    DNAKmerCounts{K}(seq)
    for _ in 1:3
        @time DNAKmerCounts{K}(seq, step)
    end
end

let
    #seq = dna"ACGTAT" ^ 1000000
    seq = first(open("chr22.fa", FASTA)).seq
    gc_enable(false)
    # K = 1
    bench_kmercounts(seq, 1, 1)
    # K = 3
    bench_kmercounts(seq, 3, 1)
    bench_kmercounts(seq, 3, 3)
    # K = 12
    bench_kmercounts(seq, 12, 1)
    bench_kmercounts(seq, 12, 3)
    bench_kmercounts(seq, 12, 6)
    bench_kmercounts(seq, 12, 12)
end
```

In summary, this pull request makes it faster for `step >= 3`, see below for more details.

master (41bb5e6d5ab98c66bd15caea4a96db78d2469744) vs. kmer (this branch):

```
~/.j/v/Bio (master|…) $ julia bencheachkmer.jl >master.txt
~/.j/v/Bio (master|…) $ git co kmer
Switched to branch 'kmer'
~/.j/v/Bio (kmer|…) $ julia bencheachkmer.jl >kmer.txt
~/.j/v/Bio (kmer|…) $ paste master.txt kmer.txt
DNAKmerCounts{1}(seq)   DNAKmerCounts{1}(seq)
step = 1        step = 1
  0.355822 seconds (12 allocations: 18.174 MB)    0.446631 seconds (12 allocations: 18.174 MB)
  0.356342 seconds (12 allocations: 18.174 MB)    0.441976 seconds (12 allocations: 18.174 MB)
  0.357583 seconds (12 allocations: 18.174 MB)    0.460076 seconds (12 allocations: 18.174 MB)
DNAKmerCounts{3}(seq)   DNAKmerCounts{3}(seq)
step = 1        step = 1
  0.339338 seconds (12 allocations: 18.175 MB)    0.488700 seconds (12 allocations: 18.175 MB)
  0.337774 seconds (12 allocations: 18.175 MB)    0.495082 seconds (12 allocations: 18.175 MB)
  0.340025 seconds (12 allocations: 18.175 MB)    0.492560 seconds (12 allocations: 18.175 MB)
DNAKmerCounts{3}(seq)   DNAKmerCounts{3}(seq)
step = 3        step = 3
  0.265743 seconds (12 allocations: 18.175 MB)    0.213258 seconds (12 allocations: 18.175 MB)
  0.260452 seconds (12 allocations: 18.175 MB)    0.210242 seconds (12 allocations: 18.175 MB)
  0.265221 seconds (12 allocations: 18.175 MB)    0.206663 seconds (12 allocations: 18.175 MB)
DNAKmerCounts{12}(seq)  DNAKmerCounts{12}(seq)
step = 1        step = 1
  1.331471 seconds (13 allocations: 82.174 MB)    1.471263 seconds (13 allocations: 82.174 MB)
  1.346299 seconds (13 allocations: 82.174 MB)    1.451716 seconds (13 allocations: 82.174 MB)
  1.338748 seconds (13 allocations: 82.174 MB)    1.474810 seconds (13 allocations: 82.174 MB)
DNAKmerCounts{12}(seq)  DNAKmerCounts{12}(seq)
step = 3        step = 3
  0.742161 seconds (13 allocations: 82.174 MB)    0.731690 seconds (13 allocations: 82.174 MB)
  0.740564 seconds (13 allocations: 82.174 MB)    0.718271 seconds (13 allocations: 82.174 MB)
  0.731268 seconds (13 allocations: 82.174 MB)    0.728188 seconds (13 allocations: 82.174 MB)
DNAKmerCounts{12}(seq)  DNAKmerCounts{12}(seq)
step = 6        step = 6
  0.684533 seconds (13 allocations: 82.174 MB)    0.426957 seconds (13 allocations: 82.174 MB)
  0.683581 seconds (13 allocations: 82.174 MB)    0.423061 seconds (13 allocations: 82.174 MB)
  0.686904 seconds (13 allocations: 82.174 MB)    0.418399 seconds (13 allocations: 82.174 MB)
DNAKmerCounts{12}(seq)  DNAKmerCounts{12}(seq)
step = 12       step = 12
  0.463767 seconds (13 allocations: 82.174 MB)    0.374275 seconds (13 allocations: 82.174 MB)
  0.461924 seconds (13 allocations: 82.174 MB)    0.367537 seconds (13 allocations: 82.174 MB)
  0.463806 seconds (13 allocations: 82.174 MB)    0.370165 seconds (13 allocations: 82.174 MB)
```

**Comment by Ward9250** [comment]
Very cool, I'll pull down the branch and try it out! I'm interested to test the other sliding window iterator and see how it compared in terms of speed.

**Comment by bicycle1885** [comment]
Thanks. There may be a simple way to improve the performance when step = 1, but at the moment I cannot find it.

**Comment by bicycle1885** [comment]
Another micro benchmark:

```julia
using Bio.Seq

function bench_kmer{T,K}(kmer::Kmer{T,K})
    a = 0
    for i in 1:K
        a += kmer[i] == DNA_A
    end
    return a
end

let
    for r in [1, 4, 8]
        x = kmer(dna"ACGT"^r)
        println(x)
        bench_kmer(x)
        n = 0
        @time for _ in 1:100000; n += bench_kmer(x); end
        @time for _ in 1:100000; n += bench_kmer(x); end
        @time for _ in 1:100000; n += bench_kmer(x); end
        @show hash(n)
    end
end
```

Result:
```
~/.j/v/Bio (master|…) $ julia benchkmer.jl >master.txt
~/.j/v/Bio (master|…) $ git co kmer
Switched to branch 'kmer'
~/.j/v/Bio (kmer|…) $ julia benchkmer.jl >kmer.txt
~/.j/v/Bio (kmer|…) $ paste master.txt kmer.txt
DNA 4-mer:      DNA 4-mer:
 ACGT   ACGT
  0.002574 seconds        0.002446 seconds
  0.002527 seconds        0.002456 seconds
  0.002525 seconds        0.002442 seconds
hash(n) = 0x7562d72720869bcf    hash(n) = 0x7562d72720869bcf
DNA 16-mer:     DNA 16-mer:
 ACGTACGTACGTACGT       ACGTACGTACGTACGT
  0.005990 seconds        0.005375 seconds
  0.006246 seconds        0.006138 seconds
  0.006786 seconds        0.007233 seconds
hash(n) = 0xeb198c897ab8a47b    hash(n) = 0xeb198c897ab8a47b
DNA 32-mer:     DNA 32-mer:
 ACGTACGTACGTACGTACGTACGTACGTACGT       ACGTACGTACGTACGTACGTACGTACGTACGT
  0.013709 seconds        0.015331 seconds
  0.012276 seconds        0.012614 seconds
  0.012287 seconds        0.012427 seconds
hash(n) = 0xcca9ac9a73f98e18    hash(n) = 0xcca9ac9a73f98e18
```

There seems to be no performance regression in this branch.

@dcjones Could you give me a comment about this change?

**Comment by Ward9250** [comment]
Looks good to me :+1:

**Comment by bicycle1885** [comment]
Thank you! Still waiting for others' opinions!

**Comment by Ward9250** [comment]
Apologies for it taking a bit of time, I know @Blahah is abroad teaching and is very busy. I haven't heard from @dcjones for a week or so. Just pinged Gitter :D

**Comment by bicycle1885** [comment]
Never mind! I don't want to pressure you at all :grin: I meant I think the ball is not in my court.

File: src/seq/eachkmer.jl [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+function each{T, K}(t::Type{Kmer{T, K}}, seq::NucleotideSequence{T}, step::Integer=1)
+    @assert K >= 0 "K must be ≥ 0 in EachKmer"
+    @assert K <= 32 "K must be ≤ 32 in EachKmer"
+    @assert step >= 1 "step must be ≥ 1"
+
+    return EachKmerIterator{T, K}(seq, step)
+end
+
+function eachkmer{T}(seq::NucleotideSequence{T}, k::Integer, step::Integer=1)
+    return each(Kmer{T, K}, seq, step)
```
Comment (dcjones) [inline src/seq/eachkmer.jl]: Should this be lowercase `k`?

Comment (Ward9250) [inline src/seq/eachkmer.jl]: Feels like it should be upper case to me.

**Comment by dcjones** [comment]
Looks good to me. Feel free to merge!

There's kind of an incongruity with k-mers being encoded differently than nucleotide sequences. That probably doesn't have much practical implication, so changing the encoding of nucleotide sequences shouldn't be a priority, but it is something we should consider doing eventually.

**Comment by Ward9250** [comment]
@dcjones  We also need to decide on the extra encoding choice of 4 bit for gaps and ambigs, or representing gaps and Multiple Sequence Alignments using the align stuff already in the making. So maybe this should be brought up in #122

File: src/seq/eachkmer.jl (hunk 1 above)
Comment (bicycle1885) [inline src/seq/eachkmer.jl]: Good catch! Fixed.

**Comment by bicycle1885** [comment]
I'm going to merge this if CI passes on v0.4.

**Comment by Ward9250** [comment]
It might say the tests fail as at present there is an issue downloading BioFmtSpecimens on the travis build, I've taken a look but haven't figured out what's wrong.

**Comment by bicycle1885** [comment]
That happened as usual; no clue to debug this problem.

**Pull request merged**